src/SmartComponents/Inventory/applications/compliance/SystemRulesTable.js

```diff
@@ -1,181 +1,364 @@
+        });
+    }
+
+    selectedRules() {
+        return this.state.rows.filter(row => row.selected).map(row => this.state.refIds[row.cells[0]]);
+    }
+
+    // This takes both the row and its children (collapsible rows) and puts them into
+    // one element. It is used only so that the rows can be compared.
+    compressRows(rows) {
+        const compressedRows = [];
+        rows.forEach((row, i) => {
+            if (row.hasOwnProperty('isOpen')) {
+                compressedRows.push({ parent: row, child: rows[i + 1] });
+            }
+        });
+
+        return compressedRows;
+    }
+
+    uncompressRows(compressedRows) {
+        const originalRows = [];
+        compressedRows.forEach((compressedRow, i) => {
+            originalRows.push(compressedRow.parent);
+            if (compressedRow.child) {
+                let child = compressedRow.child;
+                child.parent = i * 2;
+                originalRows.push(child);
+            }
         });
+        return originalRows;
     }
 
-    selectedRules = () => {
-        return this.state.rows.filter(row => row.selected).map(row => row.cells[1]);
+    onSort(_event, index, direction) {
+        // Original index is not the right column, as patternfly adds 1 because
+        // of the 'collapsible' button and 1 because of the checkbox.
+        let column = index - 2;
+        const sortedRows = this.uncompressRows(
+            this.compressRows(this.state.rows).sort(
+                (a, b) => {
+                    if (direction === SortByDirection.asc) {
+                        if (column === COMPLIANT_COLUMN) {
+                            return a.parent.cells[column].props.style.color === b.parent.cells[column].props.style.color ? 0 :
+                                a.parent.cells[column].props.style.color < b.parent.cells[column].props.style.color ? 1 : -1;
+                        } else {
+                            return a.parent.cells[column].localeCompare(b.parent.cells[column]);
+                        }
+
+                    } else {
+                        if (column === COMPLIANT_COLUMN) {
+                            return a.parent.cells[column].props.style.color === b.parent.cells[column].props.style.color ? 0 :
+                                a.parent.cells[column].props.style.color > b.parent.cells[column].props.style.color ? 1 : -1;
+                        } else {
+                            return -a.parent.cells[column].localeCompare(b.parent.cells[column]);
+                        }
+                    }
+                }
+            )
+        );
+        this.currentRows(this.state.page, this.state.itemsPerPage, sortedRows).then((currentRows) => {
+            this.setState(() => ({
+                currentRows,
+                sortBy: {
+                    index,
+                    direction
+                },
+                rows: sortedRows
+            }));
+        });
+    }
+
+    hidePassed(checked) {
```

Comment:
This function looks really complicated, can you explain what it tries to do?

Reply:
It simply checks whether the `hide passed rules` button is checked. If it is, it only displays the rules that have failed compliance, so the user can easily see what failed. 
  
To do this, I have to check the value of the compliance column: in this case, it's a component `ExclamationCircleIcon`, so I can't just check a text value and I check the color of the component. The sorting function just adheres to the JS requirements, -1 if `a` is less than `b`, 0 if they're equal, 1 if `a` is greater than `b`. It handles both sorting in ascendent and descent direction. 

After all that's done, it sets the state so that the table picks it up and displays the right arrow in the TableHeader.